### PR TITLE
Core: Pass a separate hooks context per story

### PR DESCRIPTION
**Pull request by Hypnosphi**
This fixes two kinds of issues:

1) Hooks break in Docs mode (React inline): try e.g. adding [Hooks story](https://github.com/storybookjs/storybook/compare/hooks-context?expand=1#diff-30091ae9d822fd02ee7f235bcf60f2b7) in `next` branch and interact with checkbox in Docs mode
2) Hooks break when `@storybook/client-api` is not a singleton:
> [16:33] mlegait: Hi, I've a strange issue. I've updated my Storybook to the lastest beta (I used an alpha version before). And on each story using `addon-knobs`, I've the following error Error: `Storybook preview hooks can only be called inside decorators and story functions.`. If I remove `withKnobs` I no longer have the issue. Do you have an idea?
>
> [17:43] Hypnos: Can you please create a repo with minimal reproduction?
>
> [14:18] mlegait: @Hypnos  About the issue I encountered some days ago <...>
Here is a gitHub repo where we can see the error on the `Link` story:
https://github.com/mlegait/design-system-storybook-react
>
> [02:45] Hypnos: @mlegait I ran `yarn install && yarn storybook` on you repo, and Link story opens without errors
> 
> [07:34] mlegait: Thanks @Hypnos  :smile: 
Until now I was using `npm`. I tried `npm install` and `npm clean-install `and the issue was still there. 
Then I deleted `package-lock.json` and did what you said and it works :tada: . So I assume I'll start to use `yarn`.
> 
> [14:41] Hypnos: Of course, we should work with npm as well. Looks like the problem is that `@storybook/addon-knobs` get their own copy of `@storybook/client-api`. We had a similar issue with `@storybook/addons`, so we should solve it the same way. Thanks for reporting!

### CODE REVIEW

**Comment by shilman** [comment]
@Hypnosphi Looks like this change breaks building the DLL:

```
ERROR in dll storybook_ui
Module not found: Error: Can't resolve '/tmp/storybook/lib/ui/dist/index.js' in '/tmp/storybook/lib/ui'
 @ dll storybook_ui storybook_ui[13]

ERROR in dll storybook_ui
Module not found: Error: Can't resolve '@storybook/addons' in '/tmp/storybook/lib/ui'
 @ dll storybook_ui storybook_ui[2]
error Command failed with exit code 1.
```

@ndelangen can you take a look?

**Comment by ndelangen** [comment]
I think it's the dependency cycle that's introduced

**Review by ndelangen — APPROVED** [review]

**Review by shilman — APPROVED** [review]
LGTM!

File: lib/addons/src/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,390 @@
+import { window } from 'global';
```
Comment (Hypnosphi) [inline lib/addons/src/hooks.ts]: Let's also rename the file to `preview-hooks.ts`?

**Pull request merged**